src/Firestore/Query.php

```diff
@@ -0,0 +1,477 @@
+        '>'  => self::OP_GREATER_THAN,
+        '>=' => self::OP_GREATER_THAN_OR_EQUAL,
+        '='  => self::OP_EQUAL
+    ];
+
+    private $allowedDirections = [
+        self::DIR_ASCENDING,
+        self::DIR_DESCENDING
+    ];
+
+    private $shortDirections = [
+        'ASC' => self::DIR_ASCENDING,
+        'ASCENDING' => self::DIR_ASCENDING,
+        'DESC' => self::DIR_DESCENDING,
+        'DESCENDING' => self::DIR_DESCENDING
+    ];
+
+    /**
+     * @var ConnectionInterface
+     */
+    private $connection;
+
+    /**
+     * @var ValueMapper
+     */
+    private $valueMapper;
+
+    /**
+     * @var string
+     */
+    private $parent;
+
+    /**
+     * @var array
+     */
+    private $query;
+
+    /**
+     * @var string?
+     */
+    private $transaction;
+
+    /**
+     * @param ConnectionInterface $connection A Connection to Cloud Firestore.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param array $query The Query object
+     * @param string|null The transaction ID, if the query is run in a transaction.
+     */
+    public function __construct(
+        ConnectionInterface $connection,
+        ValueMapper $valueMapper,
+        $parent,
+        array $query,
+        $transaction = null
+    ) {
+        $this->connection = $connection;
+        $this->valueMapper = $valueMapper;
+        $this->parent = $parent;
+        $this->query = $query;
+        $this->transaction = $transaction;
+
+        if (!isset($this->query['from'])) {
+            throw new \InvalidArgumentException(
+                'Cannot instantiate a query which does not specify a collection selector (`from`).'
+            );
+        }
+    }
+
+    /**
+     * Get all documents matching the provided query filters.
+     *
+     * This method will not execute an immediate service call. The Query will be
+     * executed once you begin iterating over the rows (either by iterating
+     * directly on the QuerySnapshot, or by iterating over
+     * {@see Google\Cloud\Firestore\QuerySnapshot::rows()}).
+     *
+     * Example:
+     * ```
+     * $result = $query->snapshot();
+     * ```
+     *
+     * @param array $options Configuration options.
+     * @return QuerySnapshot
+     */
+    public function snapshot(array $options = [])
+    {
+        $options['maxRetries'] = (isset($options['maxRetries']))
+            ? $options['maxReties']
+            : FirestoreClient::MAX_RETRIES;
+
+        $call = function () use ($options) {
+            unset($options['maxRetries']);
+
+            return $this->connection->runQuery([
+                'parent' => $this->parent,
+                'structuredQuery' => $this->query,
+                'transaction' => $this->transaction,
+                'retries' => 0,
+            ] + $options);
+        };
+
+        return new QuerySnapshot($this->connection, $this->valueMapper, $this, $call, $options['maxRetries']);
+    }
+
+    /**
+     * Add a SELECT to the Query.
```

Comment:
It might be good to explain that this is a projection (or explain what it does).